src/inference/AdvancedSMC.jl

```diff
@@ -17,76 +43,106 @@ must be stored in a [`TArray`](@ref) object.
-SMC(n_particles::Int, ::Tuple{}) = SMC(n_particles)
-function SMC(n_particles::Int, space::Symbol...)
-    SMC(n_particles, resample_systematic, 0.5, space)
+SMC() = SMC(resample_systematic, 0.5, ())
+SMC(::Tuple{}) = SMC()
+function SMC(space::Symbol...)
+    SMC(resample_systematic, 0.5, space)
 end
 
-function Sampler(alg::SMC, s::Selector)
-    info = Dict{Symbol, Any}()
-    info[:logevidence] = []
-    return Sampler(alg, info, s)
+mutable struct SMCState{V<:VarInfo, F<:AbstractFloat} <: AbstractSamplerState
+    vi                   ::   V
+    # The logevidence after aggregating all samples together.
+    average_logevidence  ::   F
+    particles            ::   ParticleContainer
 end
 
-function step(model, spl::Sampler{<:SMC}, vi::VarInfo)
+function SMCState(
+    model::M, 
+) where {
+    M<:Model
+}
+    vi = VarInfo(model)
     particles = ParticleContainer{Trace}(model)
-    vi.num_produce = 0;  # Reset num_produce before new sweep\.
-    set_retained_vns_del_by_spl!(vi, spl)
-    resetlogp!(vi)
 
-    push!(particles, spl.alg.n_particles, spl, vi)
+    return SMCState(vi, 0.0, particles)
+end
 
-    while consume(particles) != Val{:done}
-      ess = effectiveSampleSize(particles)
-      if ess <= spl.alg.resampler_threshold * length(particles)
-        resample!(particles,spl.alg.resampler)
-      end
-    end
+function Sampler(alg::T, model::Model, s::Selector) where T<:SMC
+    dict = Dict{Symbol, Any}()
+    state = SMCState(model)
+    return Sampler(alg, dict, s, state)
+end
 
-    ## pick a particle to be retained.
-    Ws, _ = weights(particles)
-    indx = randcat(Ws)
-    push!(spl.info[:logevidence], particles.logE)
+function sample_init!(
+    ::AbstractRNG, # Note: This function does not use the range argument.
+    model::Turing.Model,
+    spl::Sampler{<:SMC},
+    N::Integer; # Note: This function doesn't use the N argument.
+    kwargs...
+)
+    # Set the parameters to a starting value.
+    initialize_parameters!(spl; kwargs...)
 
-    return particles[indx].vi, true
-end
+    # Update the particle container now that the sampler type
+    # is defined.
+    spl.state.particles = ParticleContainer{Trace{typeof(spl),
+        typeof(spl.state.vi), typeof(model)}}(model)
 
-## wrapper for smc: run the sampler, collect results.
-function sample(model::Model, alg::SMC)
-    spl = Sampler(alg)
+    spl.state.vi.num_produce = 0;  # Reset num_produce before new sweep\.
+    set_retained_vns_del_by_spl!(spl.state.vi, spl)
+    resetlogp!(spl.state.vi)
 
-    particles = ParticleContainer{Trace}(model)
-    push!(particles, spl.alg.n_particles, spl, empty!(VarInfo(model)))
+    push!(spl.state.particles, N, spl, empty!(spl.state.vi))
 
-    while consume(particles) != Val{:done}
-      ess = effectiveSampleSize(particles)
-      if ess <= spl.alg.resampler_threshold * length(particles)
-        resample!(particles,spl.alg.resampler)
-      end
+    while consume(spl.state.particles) != Val{:done}
```

Comment:
Because `SMC` isn't meant to be run iteratively -- the particles are meant to be consumed in total. Our SMC implementation doesn't fit well into this interface design, and I think it needs to be reworked a little. In general I think the way we work with `ParticleContainers` needs a little work after this PR, since I haven't really touched it yet. @yebai and I spoke about this a bit way earlier on in the PR (I can't seem to find the comments now).

Reply:
This is a bit hacky IMO. The same trick is also used for DynamicHMC. Since this does not affect functionality, I suggest that we leave this as-is for now, and address in a separate PR.

Reply:
Sounds good to me.